io_uring_echo_server.c

```diff
@@ -189,9 +192,10 @@ void add_socket_read(struct io_uring* ring, int fd, struct iovec* iovecs, int ty
     io_uring_sqe_set_data(sqe, conn_i);
 }
 
-void add_socket_write(struct io_uring* ring, int fd, struct iovec* iovecs, int type) {
+void add_socket_write(struct io_uring* ring, int fd, size_t size, int type) {
 
     struct io_uring_sqe *sqe = io_uring_get_sqe(ring);
+    iovecs[fd].iov_len = size;
```

Comment:
For some reason this line (`iovecs[fd].iov_len = size;`) causes a ~70% performance drop:

_with_ `iovecs[fd].iov_len = size;`:

```
max@ubuntu:~/hdv/source/rust_echo_bench$ cargo run --release -- --address "localhost:6666" --number 50 --duration 30 --length 512
    Finished release [optimized] target(s) in 0.01s
     Running `target/release/echo_bench --address 'localhost:6666' --number 50 --duration 30 --length 512`
Benchmarking: localhost:6666
50 clients, running 512 bytes, 30 sec.

Speed: 387685 request/sec, 387685 response/sec
Requests: 11630554
Responses: 11630554
```

_without_ `iovecs[fd].iov_len = size;`:

```
max@ubuntu:~/hdv/source/rust_echo_bench$ cargo run --release -- --address "localhost:6666" --number 50 --duration 30 --length 512
    Finished release [optimized] target(s) in 0.01s
     Running `target/release/echo_bench --address 'localhost:6666' --number 50 --duration 30 --length 512`



Benchmarking: localhost:6666
50 clients, running 512 bytes, 30 sec.

Speed: 124698 request/sec, 124698 response/sec
Requests: 3740948
Responses: 3740947
```

Reply:
I think the lack of this line is causing to wrong behavior.

Reply:
> I think the lack of this line is causing to wrong behavior.
> 
> readv does not modify `iov_len`, which will cause writev to always send 2048 bytes.
> 
> ~Since `bufs` is uninitialized, this may cause UB.~
> 
> Ah, this seems to be expected behavior. It's marvellous, I'm interested in this reason.

It might have something to do with the used test [tool](https://github.com/haraldh/rust_echo_bench) I see the same performance jump/drop with the [epoll echo server](https://github.com/frevib/epoll-echo-server).

Reply:
You're right, `rust_echo_bench` expects to return data of same length, and return too much data will cause shorter read times.

Reply:
I also noticed that `io_uring-echo-server` did not implement full writes, which could lead to loss.

see
https://github.com/frevib/io_uring-echo-server/blob/master/io_uring_echo_server.c#L162
https://github.com/quininer/io-uring/blob/master/examples/tcp_echo.rs#L211

Reply:
> I also noticed that `io_uring-echo-server` did not implement full writes, which could lead to loss.
> 

Thanks for taking the time to look at it! What do you mean with full writes? Partial writes because of non-blocking sockets?

> see
> https://github.com/frevib/io_uring-echo-server/blob/master/io_uring_echo_server.c#L162
> https://github.com/quininer/io-uring/blob/master/examples/tcp_echo.rs#L211

Reply:
Yes, because you set `SOCK_NONBLOCK`, I think that a partial write might happen.

Reply:
Good point, thanks.

I might keep this echo server as simple as possible. The goal is to let people learn people the basics of io_uring. Maybe handling partial writes is a bit too much ;)

Reply:
Still dont know what does a partial write mean. Could you explain it further?

Also please see [my fork](https://github.com/CarterLi/io_uring-echo-server) repo, I fixed the issue there, and got some banchmark results.

Reply:
In addition, are these benchmark tools really reliable? How can you ensure that it's not the tool that limits the performance results.

Reply:
> Still don't know what does a partial write mean. Could you explain it further?
> 
> EDIT: did you mean that situations that message lengths are greater then `MAX_MESSAGE_LEN`?
> 

Partial write happens when we try to write more bytes to the socket send buffer than is actually free in this buffer. We write until the buffer is full and then return, this writing only a part of the all the bytes. If the receiving end expects the full message to return, this could lead to odd behaviour.

This happens to both blocking and non-blocking sockets. 

> Also please see [my fork](https://github.com/CarterLi/io_uring-echo-server) repo, I fixed the issue there, and got some banchmark results.

Nice, I’ll have a look asap!

Reply:
> In addition, are these benchmark tools really reliable? How can you ensure that it's not the tool that limits the performance results.

No, they are not. @quininer found out there’s a problem when returning more bytes than the amount of bytes that was received: https://github.com/frevib/io_uring-echo-server/pull/2#discussion_r376732143

This would make the epoll echo server ~70% more performant than the io_uring echo server, on my machine. This also doesn’t make any sense :)